src/full/Agda/TypeChecking/Rules/Term.hs

```diff
@@ -764,55 +764,52 @@ checkExtendedLambda ::
 checkExtendedLambda cmp i di erased qname cs e t = do
   mod <- currentModality
   if isErased erased && not (hasQuantity0 mod) then
-    genericError $ unwords
-      [ "Erased pattern-matching lambdas may only be used in erased"
-      , "contexts"
-      ]
-   else setModeUnlessInHardCompileTimeMode erased $ do
-        -- Erased pattern-matching lambdas are checked in hard
-        -- compile-time mode. For non-erased pattern-matching lambdas
-        -- run-time mode is used, unless the current mode is hard
-        -- compile-time mode.
-   -- Andreas, 2016-06-16 issue #2045
-   -- Try to get rid of unsolved size metas before we
-   -- fix the type of the extended lambda auxiliary function
-   solveSizeConstraints DontDefaultToInfty
-   lamMod <- inFreshModuleIfFreeParams currentModule  -- #2883: need a fresh module if refined params
-   t <- instantiateFull t
-   ifBlocked t (\ m t' -> postponeTypeCheckingProblem_ $ CheckExpr cmp e t') $ \ _ t -> do
-     j   <- currentOrFreshMutualBlock
-     mod <- currentModality
-     let info = setModality mod defaultArgInfo
-
-     reportSDoc "tc.term.exlam" 20 $ vcat
-       [ hsep
-         [ text $ show $ A.defAbstract di
-         , "extended lambda's implementation"
-         , doubleQuotes $ prettyTCM qname
-         , "has type:"
-         ]
-       , prettyTCM t -- <+> " where clauses: " <+> text (show cs)
-       ]
-     args     <- getContextArgs
-
-     -- Andreas, Ulf, 2016-02-02: We want to postpone type checking an extended lambda
-     -- in case the lhs checker failed due to insufficient type info for the patterns.
-     -- Issues 480, 1159, 1811.
-     abstract (A.defAbstract di) $ do
-       -- Andreas, 2013-12-28: add extendedlambda as @Function@, not as @Axiom@;
-       -- otherwise, @addClause@ in @checkFunDef'@ fails (see issue 1009).
-       addConstant qname =<< do
-         lang <- getLanguage
-         fun  <- emptyFunction
-         useTerPragma $
-           (defaultDefn info qname t lang fun)
-             { defMutual = j }
-       checkFunDef' t info (Just $ ExtLamInfo lamMod False empty) Nothing di qname $
-         List1.toList cs
-       whenNothingM (asksTC envMutualBlock) $
-         -- Andrea 10-03-2018: Should other checks be performed here too? e.g. termination/positivity/..
-         checkIApplyConfluence_ qname
-       return $ Def qname $ map Apply args
+    typeError LambdaIsErased
+  else setModeUnlessInHardCompileTimeMode erased $ do
```

Comment:
Rather than a long-ranging `if-else`, I guess the idiomatic way would be to use `when`:
```
  when (...) $ typeError ...
  setModeUnles... do
     ...<ocde>
```